(STM32) Add support for tamper event detection

GseoC opened the pull request:
The anti-tamper detection circuit is used to protect sensitive data from external attacks. On a tamper event detection, all or part of the secrets of the board are erased.

Add support for the tamper event detection with the wakeup capability and time stamping support. To be able to enable the wakeup capability, rely on the EXTI line for stm32mp13/25 SoCs instead of the GIC one. The EXTI makes the link with the GIC itself.

This series depends on: #7419

etienne-lms commented on core/drivers/stm32_tamp.c:
```diff
@@ -529,20 +924,1056 @@ static void parse_bkpregs_dt_conf(const void *fdt, int node)
+	/* Active configuration applied only if not already done. */
+	if (((io_read32(base + _TAMP_ATOR) & _TAMP_INITS) != _TAMP_INITS)) {
+		io_write32(base + _TAMP_ATCR1, atcr1);
+		FMSG("Set active conf1 %08x", atcr1);
```
Use `PRIx32`.
Ditto at lines 1303 & 1313.

GseoC commented on core/drivers/stm32_rtc.c:
```diff
@@ -543,6 +544,95 @@ static TEE_Result stm32_rtc_set_time(struct rtc *rtc, struct optee_rtc_time *tm)
+	io_setbits32(base + RTC_SCR, RTC_SCR_CTSF);
+
+	/* Overflow detection */
+	if (io_read32(base + RTC_SR) & RTC_SR_TSOVF)
```
It's not a RTC overflow. It's rather this flag is set when a timestamp event occurs while the TSF bit hasn't been cleared yet. I can add a trace but not as an error, probably a debug message.

GseoC commented on core/include/dt-bindings/tamper/st,stm32-tamp.h:
```diff
@@ -0,0 +1,25 @@
+/* Nothing for tampers 6-7*/
+#define INT_TAMPER_6				6
+#define INT_TAMPER_7				7
+#define INT_TAMPER_MONOTONIC_COUNTER_		8
```
Yes, sure, probably a bad copy paste

GseoC commented on core/drivers/stm32_tamp.c:
```diff
@@ -529,20 +924,1056 @@ static void parse_bkpregs_dt_conf(const void *fdt, int node)
+	}
+
+	if (!tamp_ext)
+		return TEE_ERROR_BAD_PARAMETERS;
```
Yes, good catch

GseoC commented on core/drivers/stm32_tamp.c:
```diff
@@ -529,20 +924,1056 @@ static void parse_bkpregs_dt_conf(const void *fdt, int node)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Find internal Tamp struct */
+	for (i = 0; i < stm32_tamp_dev->pdata.compat->int_tamp_size; i++)
```
Sure

etienne-lms commented on core/include/dt-bindings/tamper/st,stm32-tamp.h:
```diff
@@ -0,0 +1,25 @@
+#define INT_TAMPER_LSE_MONITORING		3
+#define INT_TAMPER_HSE_MONITORING		4
+#define INT_TAMPER_RTC_CALENDAR_OVERFLOW	5
+/* Nothing for tampers 6-7*/
```
space char missing before `*/`

etienne-lms commented on core/drivers/stm32_tamp.c:
```diff
@@ -1,11 +1,16 @@
-// SPDX-License-Identifier: BSD-3-Clause
+// SPDX-License-Identifier: BSD-2-Clause
```
I ack the license terms change.

etienne-lms commented on core/drivers/stm32_tamp.c:
```diff
@@ -680,19 +2157,41 @@ static TEE_Result stm32_tamp_probe(const void *fdt, int node,
+	.ext_tamp_size = ARRAY_SIZE(ext_tamp_mp25),
+	.pin_map = pin_map_mp25,
+	.pin_map_size = ARRAY_SIZE(pin_map_mp25),
 };
```
The MP13 and MP25 compat references to `int_tamp_mp*`, `ext_tamp_mp*` and `pin_map_mp*` consume useless internal RAM resident memory when pager is enabled on MP15. Could you add some `CFG_STM32MPxx` directives to prevent that?

etienne-lms replied:
Thanks

etienne-lms commented on core/drivers/stm32_tamp.c:
```diff
@@ -529,39 +924,1079 @@ static void parse_bkpregs_dt_conf(const void *fdt, int node)
+
+	if (!stm32_tamp_dev->pdata.compat)
+		return TEE_ERROR_BAD_STATE;
+
```
`assert(is_unpaged(cb))`

etienne-lms commented on core/drivers/stm32_tamp.c:
```diff
@@ -529,20 +924,1056 @@ static void parse_bkpregs_dt_conf(const void *fdt, int node)
+			return stm32_tamp_configure_ext(tamp_conf, mode, cb);
+		}
+
+	EMSG("Did not find existing tamper for ID:%u", id);
```
s/%u/%d/

GseoC commented on core/include/drivers/stm32_gpio.h:
```diff
@@ -26,12 +27,24 @@ struct stm32_pinctrl;
 }
+
+static inline unsigned int
+stm32_gpio_chip_bank_id(struct gpio_chip *chip __unused)
```
I'm surprised that I didn't come across a warning, I'll add it

GseoC commented on core/drivers/stm32_rtc.c:
```diff
@@ -543,6 +544,95 @@ static TEE_Result stm32_rtc_set_time(struct rtc *rtc, struct optee_rtc_time *tm)
+		      (dr & RTC_DR_DU_MASK);
+
+	tm->tm_mon = ((dr & RTC_DR_MT_MASK) >> RTC_DR_MT_SHIFT) * 10 +
+		     ((dr & RTC_DR_MU_MASK) >> RTC_DR_MU_SHIFT) - 1;
```
Sure, seems better

etienne-lms commented on core/drivers/stm32_tamp.c:
```diff
@@ -1826,45 +1828,44 @@ static TEE_Result stm32_configure_tamp(const void *fdt, int node)
+			continue;
+
+		if (fdt_get_status(fdt, subnode) == DT_STATUS_DISABLED)
+			continue;
```
duplicated

etienne-lms reviewed (commented):
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits 
"dts: stm32: add TAMP interrupt line for tamper events in stm32mp151.dtsi",
"dts: stm32: add passive tamper sampling configuration for stm32mp15xx-dkx",
"dts: stm32: use EXTI event instead of GIC SPI for TAMP in stm32mp131.dtsi",
"dts: stm32: add tamper event detection configuration for stm32mp135f-dk",
"dts: stm32: use EXTI event instead of GIC SPI for TAMP in stm32mp251.dtsi" and
"dts: stm32: add tamper event detection configuration for stm32mp257f-ev1".

GseoC commented:
Hi, thanks for the review.

- Comments addressed except for one in `stm32_rtc_get_timestamp()`
- Fixup squashed 
- Tags on dts files applied

etienne-lms commented on core/drivers/stm32_rtc.c:
```diff
@@ -543,6 +544,95 @@ static TEE_Result stm32_rtc_set_time(struct rtc *rtc, struct optee_rtc_time *tm)
+
+	if (IO_READ32_POLL_TIMEOUT(base + RTC_SR, value,
+				   value & RTC_SR_TSF,
+				   10, TIMEOUT_US_RTC_GENERIC))
```
I'm wondering about the time possibly spent in the interrupt routine polling on this bit.

GseoC replied:
COnsidering this is used for tamper event detection, which practically either erase platform secrets / reset the platform, I think we may accept a delay here.

etienne-lms commented:
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"drivers: stm32_gpio: add stm32_gpio_get_bank_id() helper".

GseoC commented:
Comments addressed, I will add the tag on the GPIO commit when squashing

GseoC commented on core/drivers/stm32_tamp.c:
```diff
@@ -529,39 +924,1079 @@ static void parse_bkpregs_dt_conf(const void *fdt, int node)
+	if (res != TEE_ERROR_ITEM_NOT_FOUND) {
+		active = true;
+		out_id = OUT_TAMP1 + out_tamp_id - 1;
+		if (out_tamp_id > pdata->compat->ext_tamp_size) {
```
Seems fine as IDs start at 1?

GseoC commented:
Comments addressed

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"drivers: stm32_rtc: add time stamping feature support".

`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"drivers: stm32_tamp: add tamper events detection support"

GseoC commented:
Rebased for CI

GseoC commented on core/drivers/stm32_tamp.c:
```diff
@@ -529,39 +924,1079 @@ static void parse_bkpregs_dt_conf(const void *fdt, int node)
+				io_setbits32(base + _TAMP_SCR,
+					     _TAMP_SCR_ETAMP(id));
+
+			/* TODO: reset the platform */
+			if (ret & TAMP_CB_RESET)
+				MSG("System will reset");
```
I don't currently rely on the watchdog as we return that the interrupt is handled. Nevertheless, I propose to keep as-is and I will send another patch to implement the system reset right after (I already tested it and it seems fine).

GseoC commented:
Squashed and last tags applied, thanks

GseoC commented:
Empty push for CI fail on unresponsive buildroot repo

jforissier commented on core/drivers/stm32_tamp.c:
```diff
@@ -93,12 +96,103 @@
+#define _TAMP_ATCR1_FLTEN		BIT(31)
+#define _TAMP_ATCR1_COMMON_MASK		GENMASK_32(31, 16)
+#define _TAMP_ATCR1_ETAMPAM(id)		BIT((id) - EXT_TAMP1)
+#define _TAMP_ATCR1_ATOSEL_MASK(id) \
+	GENMASK_32(((id) - EXT_TAMP1 + 1) * 2 + 7, ((id) - EXT_TAMP1) * 2 + 8)
```
```suggestion
#define _TAMP_ATCR1_ATOSEL_MASK(id) \
	({ \
		typeof(id) _id = (id); \
		GENMASK_32(((_id) - EXT_TAMP1 + 1) * 2 + 7, ((_id) - EXT_TAMP1) * 2 + 8); \
	})
```
...should get rid of the checkpatch warning (and make the macro safer to use)

GseoC commented:
Thank you for the suggestion.

I added the change at the end of the series.
I also added a minor fix to allow another prediv value for a feature => haven't dropped the tags because of how minor it is.

Thanks

GseoC commented:
@jforissier , if you're okay with it, I can squash the two patches.

jforissier commented:
@GseoC yes please.

GseoC commented:
Squashed, thank you
